### PR TITLE
Support Keycloak Dev Services for standalone OIDC Client Registration extension

### PR DESCRIPTION
**Pull request by michalvavrik**
This PR does 2 things:

- improves extendability of Keycloak Dev Service as we plan to support KC Dev Svc for Keycloak Admin Client
- starts Keycloak Dev Service when OIDC Client Registration is used without OIDC extension and OIDC Client extension (implements part of the #44143)

### CODE REVIEW

File: extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/OidcDevServicesBuildStep.java
Comment (sberyozkin) [inline extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/OidcDevServicesBuildStep.java:20]:
```suggestion
public class KeycloakDevServiceRequiredBuildStep {
```

in `quarkus-oidc`, the code which lives directly in the `.devservices.` package, is meant to support not only Keycloak in devmode, so I propose rename this build step, and move further down in the `devservices.keycloak` package. And probably makes sense to do he same  for other extensions for consistency

Comment (michalvavrik) [reply extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/devservices/OidcDevServicesBuildStep.java]: I didn't think of that, thanks.

File: extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakAdminPageBuildItem.java
Comment (sberyozkin) [inline extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakAdminPageBuildItem.java:18]: What is the idea behind the `priority` ?

Comment (michalvavrik) [reply extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakAdminPageBuildItem.java:18]: Yeah, I though we will discuss that :-D.

So far, we had one card created for Quarkus OIDC extension when Keycloak Dev Services were started. This card contained "Keycloak Provider" and "Keycloak Admin" links.

When we added support for Keycloak Dev Service started when Quarkus OIDC Client extension is used without Quarkus OIDC extension, card with "Keycloak Admin" was created for Quarkus OIDC Client. That is there are not 2 cards, always one card with "Keycloak Admin" link.

And that is what priority is for, to have just one card with "Keycloak Admin" link. I think it would be useless when for example all three OIDC, OIDC Client and OIDC Client registration were present, we had 3 cards with same link. And then imagine someone will also use Keycloak Admin Client in the future, will we show 4 cards with same link?

So, that is what priority is for, have status quo when Quarkus OIDC is present, everything will appear as it used to be one or 2 releases before (don't remember when that PR got merged).

Comment (michalvavrik) [reply extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakAdminPageBuildItem.java:18]: BTW I have tested it.

Comment (sberyozkin) [reply extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakAdminPageBuildItem.java:18]: Michal, I'm not sure we need this optimization, while it would indeed look strange to have `4/5` cards loaded, each of them having a `Keycloak Admin` link, in reality, we'll never have `4/5` cards loaded, `3` would be absolute max, `2` would be the main number IMHO, when more than 2 OIDC related extensions combine.

When I have for example `quarkus-oidc-client` and `quarkus-oidc-extension` or `quarkus-oidc-client-registration` and `quarkus-oidc-extension` loaded, I definitely would like to see 2 cards, one for `quarkus-oidc`, and another one for either  `quarkus-oidc-client` or `quarkus-oidc-client-registration`...

I don't think having a single card to represent `quarkus-oidc`, `quarkus-oidc-client`, etc when more than one OIDC or Keycloak aware extensions are loaded  would be nice...

May be we can simply agree that if `quarkus-oidc` is loaded then `Keycloak Admin` is shown only in the `quarkus-oidc` card, if `quarkus-oidc` is not loaded - then it is shown in each of the other OIDC aware cards

Comment (michalvavrik) [reply extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakAdminPageBuildItem.java:18]: When I don't have strong opinion, I just keep original behavior because it is something we are most likely to agree on. I don't really mind here, I added this priority because I thought you will want to keep how it worked. Let me drop priority altogether. Then you can test it and if it will look good to you, I am fine as well.

Comment (michalvavrik) [reply extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakAdminPageBuildItem.java]: Done. I have upgraded my Fedora OS version today and had to drop Docker for the upgrade to work. So I didn't test this last change yet, will check tomorrow when I convince Podman to work, but I expect the last commit to work just fine.

Comment (michalvavrik) [reply extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakAdminPageBuildItem.java]: I have Podman working so I checked changes. IMO it looks ok:
image

Comment (sberyozkin) [reply extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakAdminPageBuildItem.java]: Thanks @michalvavrik

File: extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java
Comment (sberyozkin) [inline extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java:90]: Does it really work though, it is a runtime property

File: extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java
Comment (sberyozkin) [inline extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java:48]: What happens if only `quarkus-oidc-client-registration` and `quarkus-oidc-client` are loaded ? Is Keycloak launched twice ?

File: extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java
Comment (michalvavrik) [reply extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java:90]: I don't have initial token, so if you don't mind, I'll answer with another question. Not sure it's fair, but:

Quarkus OIDC auth server URL property https://quarkus.io/guides/all-config#quarkus-oidc_quarkus-oidc-auth-server-url is checked before this code is executed https://github.com/quarkusio/quarkus/blob/main/extensions/devservices/keycloak/src/main/java/io/quarkus/devservices/keycloak/KeycloakDevServicesProcessor.java#L369, while this code you are pointing to is executed after KC containers has started and you are creating default realm. What is difference? Either all runtime property can be queries or none.

File: extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java
Comment (michalvavrik) [reply extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java:48]: I think I can put it even simpler: would we start KC Dev Svc for OIDC Client (/registration) if OIDC is using different authentication server? My answer is no, but if the answer is yes, we will need to tweak DEV UI because currently it doesn't support this situation.

File: extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java
Comment (michalvavrik) [reply extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java:90]: @sberyozkin I have played with a dev mode for a while now and yes, it works (though I think the answer above explain the reason why it is expect to work as well). so what happens is that when initial token is set (I did set nonsense), KC is not configured with anonymous policy and default client registration fails. You would need to configure KC with realm.json as you did inside of the OIDC client integration test.

Comment (michalvavrik) [reply extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java:90]: I also tried to apply step by step:

```
quarkus.oidc-client.enabled=false
quarkus.oidc-client-registration.enabled=false
quarkus.oidc.enabled=false
```
and enable / disable individual extensions, all works as I expected. when all 3 are disabled, KC dev service is closed. when you enable at least one of them, KC is started again.

File: extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java
Comment (sberyozkin) [inline extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java:37]: It is a bit tricky to check on my mobile phone, but it feels like some good portions of the code, for checking properties, docker, initial parts of the configurators can be pushed down to some common code for all 3 extensions... Have a look please,  LGTM overall

Comment (michalvavrik) [reply extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java]: I extracted code "duplicates" into the Keycloak Dev Service extension. It require few loops, I think I got them right, but I'll need to need to re-test it with Quarkus DEV mode and different extension combinations.

**Comment by michalvavrik** [comment]
Failures seem related, must had change logic yesterday. I'll fix it.

File: extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java
Comment (michalvavrik) [reply extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/devservices/keycloak/KeycloakDevServiceRequiredBuildStep.java]: Done. I have fixed the issue and retested it.

side note: I am starting to think I should always wait for the morning before I push changes :-)

**Review by sberyozkin — APPROVED** [review]
I've run this PR locally just to check non-Keycloak providers can still be shown in Dev UI and also did a run with the the OIDC client plus OIDc quickstart, thanks @michalvavrik

**Pull request merged**
